xcm/xcm-builder/src/weight.rs

```diff
@@ -96,3 +98,41 @@ impl<T: Get<(MultiLocation, u128)>> WeightTrader for FixedRateOfConcreteFungible
 		result
 	}
 }
+
+/// Weight trader which uses the TransactionPayment pallet to set the right price for weight and then
+/// places any weight bought into the right account.
+pub struct UsingComponents<W, A, AI, C: CurrencyT<AI>, OU>(Weight, C::Balance, PhantomData<(W, A, AI, C, OU)>);
+impl<
+	WeightToFee: WeightToFeePolynomial<Balance=Currency::Balance>,
+	AssetId: Get<MultiLocation>,
+	AccountId,
+	Currency: CurrencyT<AccountId>,
+	OnUnbalanced: OnUnbalancedT<Currency::NegativeImbalance>,
+> WeightTrader for UsingComponents<WeightToFee, AssetId, AccountId, Currency, OnUnbalanced> {
+	fn new() -> Self { Self(0, Zero::zero(), PhantomData) }
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, ()> {
```

Comment:
We'll want to communicate what went wrong.
Amount conversion failing is quite different from missing payment.